src/python/pants/backend/jvm/subsystems/zinc.py

```diff
@@ -153,6 +214,108 @@ def compiler_interface(self):
+
+    :rtype: str
+    """
+    return self._zinc_factory._scala_reflect(self._products)
+
+  @memoized_property
+  def _compiler_bridge_cache_dir(self):
+    """A directory where we can store compiled copies of the `compiler-bridge`.
+
+    The compiler-bridge is specific to each scala version.
+    Currently we compile the `compiler-bridge` only once, while bootstrapping.
+    Then, we store it in the working directory under .pants.d/zinc/<cachekey>, where
+    <cachekey> is calculated using the locations of zinc, the compiler interface,
+    and the compiler bridge.
+    """
+    hasher = sha1()
+    for cp_entry in [self.zinc, self.compiler_interface, self.compiler_bridge]:
+      hasher.update(os.path.relpath(cp_entry, get_buildroot()))
+    key = hasher.hexdigest()[:12]
+
+    return os.path.join(self._zinc_factory.get_options().pants_workdir, 'zinc', 'compiler-bridge', key)
+
+  def _make_relative(self, path):
+    """A utility function to create relative paths to the work dir"""
+    return fast_relpath(path, get_buildroot())
+
+  def _run_bootstrapper(self, bridge_jar, context):
+    bootstrapper = self._zinc_factory._compiler_bootstrapper(self._products)
+    bootstrapper_args = [
+      "--out", bridge_jar,
+      "--compiler-interface", self.compiler_interface,
+      "--compiler-bridge-src", self.compiler_bridge,
+      "--scala-compiler", self.scala_compiler,
+      "--scala-library", self.scala_library,
+      "--scala-reflect", self.scala_reflect,
+    ]
+    input_jar_snapshots = context._scheduler.capture_snapshots((PathGlobsAndRoot(
+      PathGlobs(tuple([self._make_relative(jar) for jar in bootstrapper_args[1::2]])),
+      text_type(get_buildroot())
+    ),))
+    argv = tuple(['.jdk/bin/java'] +
+                 ['-cp', bootstrapper, Zinc.ZINC_BOOTSTRAPER_MAIN] +
+                 bootstrapper_args
+    )
+    req = ExecuteProcessRequest(
+      argv=argv,
+      input_files=input_jar_snapshots[0].directory_digest,
+      output_files=(self._make_relative(bridge_jar),),
+      output_directories=(self._make_relative(self._compiler_bridge_cache_dir),),
+      description="bootstrap compiler bridge.",
+      jdk_home=self.dist.home,
+    )
+    return context.execute_process_synchronously_or_raise(req, 'zinc-subsystem', [WorkUnitLabel.COMPILER])
+
+  def compile_compiler_bridge(self, context):
+    """Compile the compiler bridge to be used by zinc, using our scala bootstrapper.
+    It will compile and cache the jar, and materialize it if not already there.
+
+    :param context: The context of the task trying to compile the bridge.
+                    This is mostly needed to use its scheduler to create digests of the relevant jars.
+    :return: The absolute path to the compiled scala-compiler-bridge jar.
+    """
+    bridge_jar = os.path.join(self._compiler_bridge_cache_dir, 'scala-compiler-bridge.jar')
+    is_executing_remotely = context.options.for_global_scope().remote_execution_server
```

Comment:
I'm not sure that this decision needs to be made explicitly here.

I think that what this method can do instead is always return a `ClasspathEntry` for the bridge jar, and have logic that looks approximately like:
```
if os.path.exists(bridge_jar):
  # Snapshot the bridge jar with `scheduler.capture_snapshot` and return a `ClasspathEntry`.
else:
  # Run `_run_bootstrapper`, materialize locally, and return a `ClasspathEntry`.
```

The only downside of this simplification is that when we're running completely remotely, we don't technically need to materialize the jar locally. But I think not having the `is_executing_remotely` logic is worth it.

Reply:
I honestly could be persuaded either way, don't really have a strong intuition about the cost of materializing.

Reply:
The overheads of either capturing or materializing a `Snapshot` shouldn't be relevant as long as it is memoized to only occur once per pants run.